edx_solutions_api_integration/mobile_api/views.py

```diff
@@ -74,3 +90,72 @@ class MobileCoursesStaticTabsDetail(MobileSecureAPIView, CoursesStaticTabsDetail
 
     def __init__(self):
         self.permission_classes += (IsStaffOrEnrolled, )
+
+
+class MobileUsersCoursesGrades(MobileListAPIView):
+    """
+    View for returning a user's course grades and the course cohort's average score.
+    """
+    def get(self, request):
+        """
+        - URI: /mobile/v1/users/courses/grades?username={username}&course_id={course_id}
+        - GET: return a JSON response of the user's course grade and cohort average
+
+        * course_id: __required__, The course ID for the course to retrieve grades for
+        * username: __optional__, A staff user can retrieve grades for different users
+        by using this parameter. A regular user can only retrieve their own grades.
+        """
+        username = request.query_params.get('username', None)
+        course_id = request.query_params.get('course_id', None)
+
+        if not course_exists(request, request.user, course_id):
+            return Response({}, status=status.HTTP_404_NOT_FOUND)
+
+        user = self.request.user
+        course_key = get_course_key(course_id)
+        if not user.is_staff and not CourseEnrollment.is_enrolled(user, course_key):
+            raise PermissionDenied
```

Comment:
During review of https://github.com/edx-solutions/api-integration/pull/67 it was pointed to me that [permission classes now exist](https://github.com/edx-solutions/api-integration/blob/master/edx_solutions_api_integration/permissions.py#L90), which do most of these checks. Can you please use them to be consistent with the other classes in this module? Sorry for not realizing this earlier.

Reply:
Also, I modified the classes a bit in https://github.com/edx-solutions/api-integration/pull/67/files, you might want to check that out as I think it will be useful in your case as well (although I'm still waiting on the reviewers' opinion).

Reply:
I am doing that now, using the `IsStaffOrEnrolled` permission class, but not `IsStaffOrOwner` as the permission checks there don't apply exactly and the new endpoint needs finer-grained checks (mainly raising 404s when the specified user, not `request.user`, is not enrolled), which to me make more sense explicitly stated in the view. 

Also I checked out master and ran the mobile_api tests and got errors from the discussion metrics test case – is that just me?

Reply:
@mtyaka Given that using those permission classes in your branch (which is now in `master`) breaks tests (at least for me), I didn't switch to using the existing permission classes yet. In the case of the grades endpoint, I don't think it's worth the effort of trying to use the existing ones as they would need customisation, making them not suitable out-of-the-box. Could write new ones, which would get used solely by this endpoint though, so not sure it's exactly worth it?

Reply:
@bogdanl That's weird, the mobile_api tests all pass for me on master. What kind of errors are you seeing?

> I am doing that now, using the `IsStaffOrEnrolled` permission class, but not `IsStaffOrOwner` as the permission checks there don't apply exactly and the new endpoint needs finer-grained checks (mainly raising 404s when the specified user, not `request.user`, is not enrolled), which to me make more sense explicitly stated in the view.

If you use both `IsStaffOrEnrolled` and `IsStaffOrOwner`, the only missing part is the check whether specified user is enrolled, right? You can still keep that check in the `get` method, but replace other checks with the two permission classes, I think.  Another other difference would be that if course with specified id does not exist, you would get a 403 instead of 404, but that's the case for all existing mobile_api endpoints, so I think 403 is fine for now for the sake of consistency, and we can fix that for all endpoints in mobile_api at a later point, if we want.

Reply:
@bogdanl Ah, I think I know why you are getting errors. You have to update https://github.com/edx-solutions/discussion-edx-platform-extensions to the latest release in your virtual environment, because only the most recent release contains a method that the new discussion metrics endpoint is using.

Reply:
@mtyaka It fails because of the new permission classes somehow, I am running the latest solutions fork master and the latest api-integration master, and running into these test failures. So weird that you don't get them?

```======================================================================
FAIL: test_with_unenrolled_user_1_True (edx_solutions_api_integration.mobile_api.tests.test_users.TestUserDiscussionMetricsApi)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/ddt.py", line 114, in wrapper
    return func(self, *args, **kwargs)
  File "/edx/app/edxapp/api-integration/edx_solutions_api_integration/mobile_api/tests/test_users.py", line 152, in test_with_unenrolled_user
    self.assertEqual(response.status_code, 200)
AssertionError: 403 != 200

======================================================================
FAIL: test_with_enrolled_user (edx_solutions_api_integration.mobile_api.tests.test_users.TestUserDiscussionMetricsApi)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/edx/app/edxapp/api-integration/edx_solutions_api_integration/mobile_api/tests/test_users.py", line 167, in test_with_enrolled_user
    self.assertEqual(response.status_code, 200)
AssertionError: 403 != 200

======================================================================
FAIL: test_with_another_users_username_1_True (edx_solutions_api_integration.mobile_api.tests.test_users.TestUserDiscussionMetricsApi)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/ddt.py", line 114, in wrapper
    return func(self, *args, **kwargs)
  File "/edx/app/edxapp/api-integration/edx_solutions_api_integration/mobile_api/tests/test_users.py", line 131, in test_with_another_users_username
    self.assertEqual(response.status_code, 200)
AssertionError: 403 != 200
```

and into similar errors when I use those classes on the Grades endpoint.

Reply:
@bogdanl That's really weird. I re-ran the tests again, they all pass for me. I also tested the endpoint with curl to verify it works correctly in different situations.

I think we need to figure out what the problem is - can you please debug to try and figure out where these 403 errors are coming from. First I would try to remove each of the two permission classes, one by one to see if it's one of them that causes the 403. If you figure out which permission class is causing the problem, you can add debug with some print or logging statements to check which condition isn't being met in the tests.